File: src/lib/voteOdds.ts
import type { VoteTally } from "../db/queries/votes";
import { priceToDecimal } from "./decimalOdds";

const MIN_SAMPLE = 3;

export type VoteOdds = {
  p_home: number;
  p_draw: number | null;
  p_away: number;
  d_home: number;
  d_draw: number | null;
  d_away: number;
  total: number;
  lowSample: boolean;
};

/**
 * Stake-weighted crowd implied probability: each outcome's share of total
 * bottles wagered (betting more on an outcome lowers its odds, diluting your
 * own multiplier). Laplace-smoothed by 1 bottle so a zero-stake outcome still
 * yields a finite decimal. This is the settlement basis.
 */
export function computeVoteOdds(
  tally: VoteTally,
  allowsDraw: boolean,
): VoteOdds | null {
  if (tally.stakeTotal === 0) return null;

  const outcomes = allowsDraw ? 3 : 2;
  const smoothedTotal = tally.stakeTotal + outcomes;
  const pHome = (tally.home + 1) / smoothedTotal;
  const pAway = (tally.away + 1) / smoothedTotal;
  const pDraw = allowsDraw ? (tally.draw + 1) / smoothedTotal : null;

  return {
    p_home: pHome,
    p_draw: pDraw,
    p_away: pAway,
    d_home: priceToDecimal(pHome),
    d_draw: pDraw == null ? null : priceToDecimal(pDraw),
    d_away: priceToDecimal(pAway),
    total: tally.voters,
    lowSample: tally.voters < MIN_SAMPLE,
  };
}

File: rails/app/models/vote_odds.rb
# Pari-mutuel pool odds from the crowd's stakes. An outcome's implied
# probability is its share of all bottles wagered; its decimal odds are the
# whole pool divided by that outcome's stake — i.e. what each backing bottle
# returns if it wins. An outcome with no stake has no defined odds (nil).
#
# Mirrors settlement exactly: losers forfeit their stake into the pool and
# winners split it in proportion to stake, so payouts can never exceed the pool.
class VoteOdds
  MIN_SAMPLE = 3
  # Min gap (percentage points) between market and crowd before surfacing a lead.
  LEAD_DIVERGENCE_PCT = 33

  attr_reader :p_home, :p_draw, :p_away, :d_home, :d_draw, :d_away, :total, :low_sample

  # `tally` responds to home / draw / away (stake sums), stake_total and voters.
  # Returns nil when nothing has been wagered (no defined odds).
  def self.from_tally(tally, allows_draw:)
    return nil if tally.stake_total.zero?

    new(tally, allows_draw)
  end

  def initialize(tally, allows_draw)
    @p_home = share(tally.home, tally.stake_total)
    @p_draw = allows_draw ? share(tally.draw, tally.stake_total) : nil
    @p_away = share(tally.away, tally.stake_total)
    @d_home = decimal(tally.home, tally.stake_total)
    @d_draw = allows_draw ? decimal(tally.draw, tally.stake_total) : nil
    @d_away = decimal(tally.away, tally.stake_total)
    @total = tally.voters
    @low_sample = tally.voters < MIN_SAMPLE
  end

  def low_sample?
    @low_sample
  end

  private

  def share(stake, stake_total)
    stake / stake_total
  end

  def decimal(stake, stake_total)
    stake.positive? ? stake_total / stake : nil
  end
end
